### PR TITLE
Access table associations using the dot syntax

### PR DESCRIPTION
**Pull request by sdustinh**
Reference: #11366 #11369

### CODE REVIEW

**Comment by ADmad** [comment]
Instead of modifying `associations()` wouldn't it be better to add a new `getAssociations()` method which supports the new dot syntax and deprecate the former? Two birds with one stone.

**Comment by sdustinh** [comment]
:+1: That would be awesome. I'll add it.

File: src/ORM/Table.php [hunk 1]
```diff
@@ -872,14 +872,27 @@ public function hasBehavior($name)
...
+     * $users = $this->association('Articles.Comments.Users');
+     * ```
      *
      * @param string $name the alias used for the association.
      * @return \Cake\ORM\Association|null Either the association or null.
      */
     public function association($name)
     {
-        return $this->_associations->get($name);
+        list($name, $next) = array_pad(explode('.', $name, 2), 2, null);
```
Comment (ADmad) [inline src/ORM/Table.php]: Optimization should be added for checking whether name contains `.` and return early from the method.

**Review by ADmad — CHANGES_REQUESTED** [review]

**Comment by dereuromark** [comment]
But singular association() => getAssociation() afaik.

**Comment by ADmad** [comment]
> But singular association() => getAssociation() afaik.

Yeah, sorry I typoed in previous comment.

File: tests/TestCase/ORM/TableTest.php [hunk 2]
```diff
@@ -511,6 +511,30 @@ function (Event $event, $query, $options) use ($expected) {
...
+        $members = TableRegistry::get('Members');
+        $groupsMembers = TableRegistry::get('GroupsMembers');
+
+        $groups->belongsToMany('Members');
+        $groups->hasMany('GroupsMembers');
+        $groupsMembers->belongsTo('Members');
+        $members->belongsToMany('Groups');
+
+        $association = $groups->association('GroupsMembers.Members.Groups');
+        $this->assertInstanceOf('Cake\ORM\Association\BelongsToMany', $association);
```
Comment (markstory) [inline tests/TestCase/ORM/TableTest.php]: You could use `BelongsToMany::class` here 🏌️

**Comment by sdustinh** [comment]
@ADmad This should be good to go unless you guys see anything you want changed.

File: src/ORM/Table.php [hunk 3]
```diff
@@ -874,11 +874,40 @@ public function hasBehavior($name)
...
+     * ```
+     * $users = $this->getAssociation('Articles.Comments.Users');
+     * ```
+     *
+     * @param string $name the alias used for the association.
+     * @return \Cake\ORM\Association|null Either the association or null.
+     */
+    public function getAssociation($name)
+    {
+        if (strpos($name, '.') !== false) {
```
Comment (ADmad) [inline src/ORM/Table.php]: We normally follow "return early" paradigm so this should instead check for false and return the association.

Comment (sdustinh) [reply src/ORM/Table.php]: No problem. I'll take care of it now.

File: src/ORM/Table.php [hunk 4]
```diff
@@ -874,12 +874,41 @@ public function hasBehavior($name)
...
+    {
+        if (strpos($name, '.') === false) {
+            return $this->_associations->get($name);
+        }
+
+        list($name, $next) = array_pad(explode('.', $name, 2), 2, null);
+        $result = $this->_associations->get($name);
+
+        if ($result !== null && $next !== null) {
+            $result = $result->getTarget()->association($next);
```
Comment (ADmad) [inline src/ORM/Table.php]: You are using deprecated `association()` method :smile:

Comment (sdustinh) [reply src/ORM/Table.php]: Haha, jesus.

File: src/ORM/Table.php [hunk 5]
```diff
@@ -874,12 +874,41 @@ public function hasBehavior($name)
...
      * Returns an association object configured for the specified alias if any
      *
+     * @deprecated 3.6.0 Use getAssociation() instead.
      * @param string $name the alias used for the association.
      * @return \Cake\ORM\Association|null Either the association or null.
      */
     public function association($name)
     {
-        return $this->_associations->get($name);
+        return $this->getAssociation($name);
```
Comment (markstory) [inline src/ORM/Table.php]: Could we get a `deprecationWarning()` here? It'll save some time in the future.

**Comment by sdustinh** [comment]
@markstory Added a deprecation warning. I wasn't sure if you guy were adding tests for the deprecation warning or not. If you want that, I would be happy to add that as well.

**Comment by markstory** [comment]
>  I wasn't sure if you guy were adding tests for the deprecation warning or not. If you want that, I would be happy to add that as well.

We do have tests for deprecated methods, but we're also in the process of updating all the internal usage of deprecated methods so that they are unused.

**Comment by sdustinh** [comment]
@markstory Would you like me to replace all instances of `Table::association()` with `Table::getAssociation()` in this PR as well? If so, should this include those used in the tests?

**Comment by markstory** [comment]
@sdustinh That would be ideal. All uses in tests should be updated except one that is a test for the association() method to help prevent regressions until 4.0

**Comment by sdustinh** [comment]
@markstory I replaced all instances of `Table::association()` with `Table::getAssociation()` in both `src/` and `tests/`. I wasn't able to find a test specific for `Table::association()`. For now, I'll just push up those changes for review and if there is a test in there that I missed I'll patch that after.

**Comment by markstory** [comment]
Looking good :+1:

**Pull request merged**